lib/documents.js

```diff
@@ -2263,12 +2264,143 @@ function writeAllDocumentsImpl(options) {
   return inputStream;
 }
 
+function finishWriter(operation) {
+  operation.jobState.requesterCount--;
+
+  if(operation.jobState.requesterCount<1) {
+
+    if(operation.jobState.jobOptions && operation.jobState.jobOptions.onCompletion){
+
+      let summary = ('Number of documents written successfully = '+
+          ((operation.jobState.docsWrittenSuccessfully>0)?operation.jobState.docsWrittenSuccessfully:0)+'\n'+
+          'Number of documents failed to be written = '+
+          ((operation.jobState.docsFailedToBeWritten>0)?operation.jobState.docsFailedToBeWritten:0)
+      );
+      if(operation.jobState.error){
+        summary = summary + 'Processing was cancelled by an error - '+operation.error.toString();
+      }
+      operation.jobState.jobOptions.onCompletion(summary);
+    }
+
+  }
+}
+
+function onWriteAllInit(output) {
+
+  let maxRequesters = output.length;
+  const operation = this;
+  const jobState = operation.jobState;
+
+  if(jobState.jobOptions && jobState.jobOptions.concurrentRequests) {
+    if(jobState.jobOptions.concurrentRequests.multipleOf &&
+        jobState.jobOptions.concurrentRequests.multipleOf === 'hosts'){
+      const hosts = new Set();
+      for(let i=0; i<output.length; i++) {
+        hosts.add(output[i].host);
+      }
+      maxRequesters = hosts.size;
+    }
+    const multiplier = (jobState.jobOptions.concurrentRequests.multiplier &&
+        jobState.jobOptions.concurrentRequests.multiplier>0)?
+        jobState.jobOptions.concurrentRequests.multiplier:0;
+    maxRequesters = (multiplier >0) ? (maxRequesters*multiplier):maxRequesters;
+
+  }
+
+  for (let i = 0; i < maxRequesters; i++) {
+    jobState.requesterCount++;
+    onWriteAllDocs(operation,i);
+  }
+
+}
+
+function onWriteAllDocs(operation, writerId) {
+
+  const writeBatchArray = [];
+  const jobState = operation.jobState;
+  const batchSize = (jobState.jobOptions && jobState.jobOptions.batchSize &&
+      jobState.jobOptions.batchSize >0) ?
+      jobState.jobOptions.batchSize:100;
+
+  a:
+      for (let i = 0; i < batchSize; i++) {
+        const record = jobState.stream.read();
+
+        if(jobState.jobOptions && jobState.jobOptions.defaultMetadata) {
+          writeBatchArray.push(jobState.jobOptions.defaultMetadata);
+        }
+        if (record === null) {
+          switch (writeBatchArray.length) {
+            case 0: {
+              finishWriter(operation);
+              break a;
+            }
+            default: {
+              writeDocumentsImpl.call(jobState.docInstance, false, writeBatchArray)
+                  .result((out) =>{
+                    jobState.docsWrittenSuccessfully+= writeBatchArray.length;
+                      finishWriter(operation);
+                      })
+                  .catch((err) => {
+                    jobState.docsFailedToBeWritten+= writeBatchArray.length;
+                    jobState.error = err;
+                    jobState.stream.emit('error', err);
+                  });
```

Comment:
Given that the call to and result processing for writeDocumentsImpl() is the same on line 2357 and 2375 except for the last call in the success and error handlers, maybe it's time to eliminate the duplicate code.  

One possibility would be for the default case to set a flag that can be tested to guard the call to finishWriter() after the call to onBatchSuccess() -- in other words, after current line 2385